In [8]:
import pandas as pd
import statsmodels.api as sm

In [2]:
toyota = pd.read_csv('ToyotaCorolla.csv')

In [3]:
toyota.head()

Price  Age     KM Fuel_Type  HP  MetColor  Automatic  Doors
0  13500   23  46986    Diesel  90         1          0      3
1  13750   23  72937    Diesel  90         1          0      3
2  13950   24  41711    Diesel  90         1          0      3
3  14950   26  48000    Diesel  90         0          0      3
4  13750   30  38500    Diesel  90         0          0      3

In [4]:
toyota.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1436 entries, 0 to 1435
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Price      1436 non-null   int64 
 1   Age        1436 non-null   int64 
 2   KM         1436 non-null   int64 
 3   Fuel_Type  1436 non-null   object
 4   HP         1436 non-null   int64 
 5   MetColor   1436 non-null   int64 
 6   Automatic  1436 non-null   int64 
 7   Doors      1436 non-null   int64 
dtypes: int64(7), object(1)
memory usage: 89.9+ KB


Since fuel_type is object, we cannot consider as dummy variables; We need to convert to int and then use it as dummy variables

In [6]:
toyota.describe().T

count          mean           std     min      25%      50%  \
Price      1436.0  10730.824513   3626.964585  4350.0   8450.0   9900.0   
Age        1436.0     55.947075     18.599988     1.0     44.0     61.0   
KM         1436.0  68533.259749  37506.448872     1.0  43000.0  63389.5   
HP         1436.0    101.502089     14.981080    69.0     90.0    110.0   
MetColor   1436.0      0.674791      0.468616     0.0      0.0      1.0   
Automatic  1436.0      0.055710      0.229441     0.0      0.0      0.0   
Doors      1436.0      4.033426      0.952677     2.0      3.0      4.0   

                75%       max  
Price      11950.00   32500.0  
Age           70.00      80.0  
KM         87020.75  243000.0  
HP           110.00     192.0  
MetColor       1.00       1.0  
Automatic      0.00       1.0  
Doors          5.00       5.0

In [12]:
y = toyota['Price']
X = toyota[['Age','KM','HP','MetColor','Automatic']]

X_with_intercept = sm.add_constant(X)

# Fit the linear regression model
model = sm.OLS(y, X_with_intercept)
results = model.fit()

In [13]:
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Price   R-squared:                       0.813
Model:                            OLS   Adj. R-squared:                  0.812
Method:                 Least Squares   F-statistic:                     1240.
Date:                Mon, 12 Feb 2024   Prob (F-statistic):               0.00
Time:                        15:54:58   Log-Likelihood:                -12605.
No. Observations:                1436   AIC:                         2.522e+04
Df Residuals:                    1430   BIC:                         2.525e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       1.637e+04    355.781     45.998      0.000    1.57e+04    1.71e+04
Age         -155.4339      2.605    -59.658      0.000    -160.545    -150.323
KM            -0.0108      0.001     -7.994      0.000      -0.013      -0.008
HP            36.7672      2.943     12.494      0.000      30.994      42.540
MetColor      44.0261     89.252      0.493      0.622    -131.052     219.105
Automatic    747.9693    182.321      4.102      0.000     390.325    1105.614
==============================================================================
Omnibus:                      430.424   Durbin-Watson:                   1.349
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3263.789
Skew:                           1.183   Prob(JB):                         0.00
Kurtosis:                       9.996   Cond. No.                     6.70e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 6.7e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

For the two dummy variables:
Metcolor : Comparing to cars with no metallic color, cars with it sells 44 dollars more, holding evrything else constant. But its not significant.

Automatic: Comparing to cars with no auto transmission, cars with it sells for 748 dollars more, holding everything else constant

In [26]:
toyota['Fuel_Type'].unique()

array(['Diesel', 'Petrol', 'CNG'], dtype=object)

In [27]:
#Change the variable data type from integer to categorical
toyota['Doors'].unique()
toyota['Doors'] = toyota['Doors'].astype('category')

In [28]:
toyota.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1436 entries, 0 to 1435
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype   
---  ------     --------------  -----   
 0   Price      1436 non-null   int64   
 1   Age        1436 non-null   int64   
 2   KM         1436 non-null   int64   
 3   Fuel_Type  1436 non-null   object  
 4   HP         1436 non-null   int64   
 5   MetColor   1436 non-null   int64   
 6   Automatic  1436 non-null   int64   
 7   Doors      1436 non-null   category
dtypes: category(1), int64(6), object(1)
memory usage: 80.3+ KB


In [35]:
#Generate dummy variables
X_car_1 = toyota[['Age','KM','HP','MetColor','Automatic','Doors','Fuel_Type']]
X_car_1 = pd.get_dummies(X_car_1, columns = ['Doors','Fuel_Type'], drop_first=True)

In [36]:
X_car_1.head()

Age     KM  HP  MetColor  Automatic  Doors_3  Doors_4  Doors_5  \
0   23  46986  90         1          0        1        0        0   
1   23  72937  90         1          0        1        0        0   
2   24  41711  90         1          0        1        0        0   
3   26  48000  90         0          0        1        0        0   
4   30  38500  90         0          0        1        0        0   

   Fuel_Type_Diesel  Fuel_Type_Petrol  
0                 1                 0  
1                 1                 0  
2                 1                 0  
3                 1                 0  
4                 1                 0

In [39]:
X_car_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1436 entries, 0 to 1435
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   Age               1436 non-null   int64
 1   KM                1436 non-null   int64
 2   HP                1436 non-null   int64
 3   MetColor          1436 non-null   int64
 4   Automatic         1436 non-null   int64
 5   Doors_3           1436 non-null   uint8
 6   Doors_4           1436 non-null   uint8
 7   Doors_5           1436 non-null   uint8
 8   Fuel_Type_Diesel  1436 non-null   uint8
 9   Fuel_Type_Petrol  1436 non-null   uint8
dtypes: int64(5), uint8(5)
memory usage: 63.2 KB


In [37]:
X_with_intercept_car1 = sm.add_constant(X_car_1)

# Fit the linear regression model
model_car = sm.OLS(y, X_with_intercept_car1)
results_car = model_car.fit()

In [38]:
results_car.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Price   R-squared:                       0.837
Model:                            OLS   Adj. R-squared:                  0.835
Method:                 Least Squares   F-statistic:                     729.4
Date:                Mon, 12 Feb 2024   Prob (F-statistic):               0.00
Time:                        16:28:19   Log-Likelihood:                -12506.
No. Observations:                1436   AIC:                         2.503e+04
Df Residuals:                    1425   BIC:                         2.509e+04
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
const             1.342e+04   1170.837     11.461      0.000    1.11e+04    1.57e+04
Age               -139.0234      2.726    -50.993      0.000    -144.371    -133.675
KM                  -0.0192      0.001    -13.197      0.000      -0.022      -0.016
HP                  57.2648      3.188     17.963      0.000      51.011      63.518
MetColor            19.6851     83.769      0.235      0.814    -144.639     184.009
Automatic          880.6240    171.086      5.147      0.000     545.017    1216.231
Doors_3           -411.5592   1044.060     -0.394      0.693   -2459.619    1636.501
Doors_4           -540.4785   1050.152     -0.515      0.607   -2600.488    1519.531
Doors_5            -51.7080   1044.402     -0.050      0.961   -2100.439    1997.023
Fuel_Type_Diesel  2805.4187    390.330      7.187      0.000    2039.735    3571.102
Fuel_Type_Petrol   549.3283    368.973      1.489      0.137    -174.461    1273.118
==============================================================================
Omnibus:                      214.226   Durbin-Watson:                   1.376
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1314.772
Skew:                           0.531   Prob(JB):                    3.17e-286
Kurtosis:                       7.566   Cond. No.                     4.23e+06
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.23e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""